cli/azd/pkg/tools/docker/docker.go

```diff
@@ -65,27 +66,70 @@ func (d *Docker) versionInfo() tools.VersionInfo {
 
-	// Skip leading zeroes to allow inexact parsing for version formats that are not truly SemVer compliant.
-	// Example: docker has versions like 17.09.0 (non semver) instead of 17.9.0 (semver)
-	versionSplit := strings.Split(ver, ".")
-	for key, val := range versionSplit {
-		verInt, err := strconv.Atoi(val)
+// dockerVersionReleaseBuildRegexp is a regular expression which matches the three part version number
+// from a docker version from an official release. The major and minor components are captured.
+var dockerVersionReleaseBuildRegexp = regexp.MustCompile(`^(\d+).(\d+).\d+`)
+
+// dockerVersionMasterBuildRegexp is a regular expression which matches the three part version number
+// from a docker version from a release from master. Each version component is captured independently.
+var dockerVersionMasterBuildRegexp = regexp.MustCompile(`^master-dockerproject-(\d+)-(\d+)-(\d+)`)
+
+// isSupportedDockerVersion returns true if the version string appears to be for a docker version
+// of 17.09 or later and false if it does not.
+func isSupportedDockerVersion(cliOutput string) (bool, error) {
+	log.Printf("determining version from docker --version string: %s", cliOutput)
+
+	matches := dockerVersionStringRegexp.FindStringSubmatch(cliOutput)
+
+	// (3 matches, the entire string, and the two captures)
+	if len(matches) != 3 {
+		return false, fmt.Errorf("could not extract version component from docker version string")
+	}
+
+	version := matches[1]
+	build := matches[2]
+
+	log.Printf("extracted docker version: %s, build: %s from version string", version, build)
+
+	// For official release builds, the version number looks something like:
+	//
+	// 17.09.0-ce or 20.10.17+azure-1
+	//
+	// Note this is not a semver (the leading zero in the second component of the 17.09 string is not allowed per semver)
+	// so we need to take this apart ourselves.
+	if releaseVersionMatches := dockerVersionReleaseBuildRegexp.FindStringSubmatch(version); releaseVersionMatches != nil {
+		major, err := strconv.Atoi(releaseVersionMatches[1])
+		if err != nil {
+			return false, fmt.Errorf("failed to convert major version component %s to an integer: %w", releaseVersionMatches[1], err)
+		}
+
+		minor, err := strconv.Atoi(releaseVersionMatches[2])
 		if err != nil {
-			return semver.Version{}, err
+			return false, fmt.Errorf("failed to convert minor version component %s to an integer: %w", releaseVersionMatches[2], err)
 		}
-		versionSplit[key] = strconv.Itoa(verInt)
+
+		return (major > 17 || (major == 17 && minor >= 9)), nil
 	}
 
-	semver, err := semver.Parse(strings.Join(versionSplit, "."))
-	if err != nil {
-		return semver, err
+	// For builds which come out of master, we'll assume any build from 2018 or later will work
+	// (since we support 17.09 which was released in September of 2017)
+	if masterVersionMatches := dockerVersionMasterBuildRegexp.FindStringSubmatch(version); masterVersionMatches != nil {
+		year, err := strconv.Atoi(masterVersionMatches[1])
```

Comment:
@hemarina has an idea here that we could instead translate the `YYYY-MM-DD` part to `YYYY.MM.DD` and then parse that as a semver.  I think if we did that we'd at least want to subtract 2000 from the version, so we ended up with something like YY.MM.DD, but I also am not 100% sure exactly what the right cutoff is here.  So I figured anything newer 2017 would be good enough for us.